From 496e92ef4e00cbb34e1b18394f0d993eb233b764 Mon Sep 17 00:00:00 2001
From: Dimitri John Ledkov <dimitri.ledkov@surgut.co.uk>
Date: Wed, 16 Apr 2025 12:36:33 +0100
Subject: [PATCH 4/4] lddtree: support alternative LD_LIBRARY_PATH called
 AUDITWHEEL_LD_LIBRARY_PATH

Sometimes wheels are different only just by shared libraries that are
vendored in. Sometimes the wheel being repaired and the libraries that
need to be vendored in are accessible on the host that was not used
for building the wheel. In such cases, it could be the case that
shared libraries used by python that is executing the auditwheel
script, are incompatible (too old or too new) than those that need to
be vendored.

In such cases, LD_LIBRARY_PATH is not convenient to use, as the python
interpreter for the auditwheel script is crashing.

Add an AUDITWHEEL_LD_LIBRARY_PATH whichis used by lddtree, but
otherwise does not affect the python interpreter executing auditwheel
script.

This helps vendoring in older libraries from an alternative path, into
a wheel built against an older python, whilst otherwise running
auditwheel repair on a modern system with a much newer python. This is
particularly useful when stripping a wheel of vendored libraries, and
repair it again to update the shared library dependencies coming from
an unpacked chroot.

Separately it also helps creating alternative wheels with shared
libraries rebuilt with higher march settings. Given that python
upstream doesn't support loading optimised libraries as has been
implemented and supported in Intel ClearLinux Python which allows one
to have alternative march setting shared library deps.
---
 src/auditwheel/lddtree.py                | 13 +++++++++++--
 tests/integration/test_bundled_wheels.py |  6 ++++++
 2 files changed, 17 insertions(+), 2 deletions(-)

diff --git a/src/auditwheel/lddtree.py b/src/auditwheel/lddtree.py
index 44785791..661446a4 100644
--- a/src/auditwheel/lddtree.py
+++ b/src/auditwheel/lddtree.py
@@ -321,8 +321,17 @@ def load_ld_paths(
     """
     ldpaths: dict[str, list[str]] = {"conf": [], "env": [], "interp": []}
 
-    # Load up $LD_LIBRARY_PATH.
-    env_ldpath = os.environ.get("LD_LIBRARY_PATH")
+    # Load up $AUDITWHEEL_LD_LIBRARY_PATH and $LD_LIBRARY_PATH
+    env_ldpath = ":".join(
+        filter(
+            None,
+            (
+                os.environ.get("AUDITWHEEL_LD_LIBRARY_PATH"),
+                os.environ.get("LD_LIBRARY_PATH"),
+            ),
+        )
+    )
+
     if env_ldpath is not None:
         if root != "/":
             log.warning("ignoring LD_LIBRARY_PATH due to ROOT usage")
diff --git a/tests/integration/test_bundled_wheels.py b/tests/integration/test_bundled_wheels.py
index 7fa01700..97f12487 100644
--- a/tests/integration/test_bundled_wheels.py
+++ b/tests/integration/test_bundled_wheels.py
@@ -60,6 +60,12 @@
             frozenset([f"{HERE}/*"]),
             "LD_LIBRARY_PATH",
         ),
+        (
+            "cffi-1.5.0-cp27-none-linux_x86_64.whl",
+            {"libpython2.7.so.1.0"},
+            frozenset([f"{HERE}/*"]),
+            "AUDITWHEEL_LD_LIBRARY_PATH",
+        ),
         (
             "cffi-1.5.0-cp27-none-linux_x86_64.whl",
             {"libffi.so.5", "libpython2.7.so.1.0"},
